### PR TITLE
Fix and improvements to ImageStack 

### LINKED ISSUE #5959: ImageStack doesn't allow non-square data
#### ALL software version info
holoviews 1.18.0
bokeh 3.3.0

#### Description of expected behavior and the observed behavior
Looks like the x/y dimensions got swapped somewhere.

```python
hv.ImageStack(np.random.random((10,10,3))) # works
hv.ImageStack(np.random.random((10,15,3))) # breaks
```

gives the following traceback

<details>

```
---------------------------------------------------------------------------
DataError                                 Traceback (most recent call last)
Cell In[107], line 1
----> 1 hv.ImageStack(np.random.random((10,15,3)))

File ~/micromamba/envs/image_analysis/lib/python3.11/site-packages/holoviews/element/raster.py:547, in ImageStack.__init__(self, data, kdims, vdims, **params)
    545     elif isinstance(data, dict):
    546         vdims = [Dimension(key) for key in data.keys() if key not in self.kdims]
--> 547 super().__init__(data, kdims=kdims, vdims=vdims, **params)

File ~/micromamba/envs/image_analysis/lib/python3.11/site-packages/holoviews/element/raster.py:279, in Image.__init__(self, data, kdims, vdims, bounds, extents, xdensity, ydensity, rtol, **params)
    276 else:
    277     params['rtol'] = config.image_rtol
--> 279 Dataset.__init__(self, data, kdims=kdims, vdims=vdims, extents=extents, **params)
    280 if not self.interface.gridded:
    281     raise DataError("{} type expects gridded data, {} is columnar. "
    282                     "To display columnar data as gridded use the HeatMap "
    283                     "element or aggregate the data (e.g. using rasterize "
    284                     "or np.histogram2d).".format(type(self).__name__, self.interface.__name__))

File ~/micromamba/envs/image_analysis/lib/python3.11/site-packages/holoviews/core/data/__init__.py:329, in Dataset.__init__(self, data, kdims, vdims, **kwargs)
    326 kdims, vdims = kwargs.get('kdims'), kwargs.get('vdims')
    328 validate_vdims = kwargs.pop('_validate_vdims', True)
--> 329 initialized = Interface.initialize(type(self), data, kdims, vdims,
    330                                    datatype=kwargs.get('datatype'))
    331 (data, self.interface, dims, extra_kws) = initialized
    332 super().__init__(data, **dict(kwargs, **dict(dims, **extra_kws)))

File ~/micromamba/envs/image_analysis/lib/python3.11/site-packages/holoviews/core/data/interface.py:253, in Interface.initialize(cls, eltype, data, kdims, vdims, datatype)
    251     continue
    252 try:
--> 253     (data, dims, extra_kws) = interface.init(eltype, data, kdims, vdims)
    254     break
    255 except DataError:

File ~/micromamba/envs/image_analysis/lib/python3.11/site-packages/holoviews/core/data/grid.py:134, in GridInterface.init(cls, eltype, data, kdims, vdims)
    129         raise error('The shape of the %s value array does not '
    130                     'match the expected dimensionality indicated '
    131                     'by the key dimensions. Expected %d-D array, '
    132                     'found %d-D array.' % (vdim, len(expected), len(shape)))
    133     elif any((e not in (s, s + 1)) for s, e in zip(shape, valid_shape)):
--> 134         raise error(f'Key dimension values and value array {vdim} '
    135                     f'shapes do not match. Expected shape {valid_shape}, '
    136                     f'actual shape: {shape}', cls)
    137 return data, {'kdims':kdims, 'vdims':vdims}, {}

DataError: Key dimension values and value array level_0 shapes do not match. Expected shape (15, 10), actual shape: (10, 15)

GridInterface expects gridded data, for more information on supported datatypes see http://holoviews.org/user_guide/Gridded_Datasets.html
````

</details>

### LINKED ISSUE #5960: ImageStack docstring is wrong
#### ALL software version info
holoviews 1.18.0
bokeh 3.3.0

#### Description of expected behavior and the observed behavior
`hv.ImageStack`'s docstring is:
```
Type of data inputs:
    - 3D ndarray (x,y,level)
    - A list of 2D ndarrays
    - A dict of 2D ndarrays (key=level: value=2D ndarray)
    - xarray with all the whistles
```

Only the first works.

```python
hv.ImageStack(np.random.random((10,10,3))) # works
hv.ImageStack([np.random.random((10,10)) for i in range(3)]) # breaks
hv.ImageStack({i: np.random.random((10,10)) for i in range(3)}) # breaks
# I didn't bother testing xarray input
```

### PR DESCRIPTION
**Pull request by Hoxbro**
Fixes #5959 and fixes #5960 

---
TODO:
- [x] <s> Think about the transpose problem </s> The main problem is that a 2D Numpy array shape is defined as (y, x), **_not_** (x, y)
- [x] Add tests
- [x] Improve the docstring of `hv.ImageStack`
---

``` python
import holoviews as hv
import numpy as np
import panel as pn
import xarray as xr


hv.extension("bokeh")


p0 = hv.ImageStack(
    (np.arange(10), np.arange(15), np.arange(150).reshape(10, 15, 1))
).opts(num_colors=150, width=500)


p1 = hv.ImageStack(
    (np.arange(20, 30), np.arange(10, 20), np.arange(100).reshape(10, 10, 1))
).opts(num_colors=100, width=500)


p2 = hv.ImageStack(np.arange(100).reshape(10, 10, 1)).opts(num_colors=100, width=500)


da = xr.DataArray(coords={"x": np.arange(10), "y": np.arange(15)}, dims=["x", "y"])

da_1 = da.copy()
da_1.name = "a"
da_1.values = np.arange(150).reshape(10, 15)

ds = xr.merge([da_1])

p3 = hv.ImageStack(ds).opts(
    num_colors=140,
    width=500,
)

p = (p0 + p1 + p2 + p3).opts(shared_axes=False)

pn.panel(p).servable()
```

image

### CODE REVIEW

**Comment by Hoxbro** [comment]
Added conversion of different inputs:
``` python
import holoviews as hv
import numpy as np

hv.extension("bokeh")

np.random.seed(42)
data = np.random.random((10, 15, 3))

p0 = hv.ImageStack(data)
p1 = hv.ImageStack([data[:, :, i] for i in range(3)])
p2 = hv.ImageStack({str(i): data[:, :, i] for i in range(3)})
p0 + p1 + p2
```

image

**Comment by Hoxbro** [comment]
I'm not sure about the transposing, needs to think more about it.

With the following code:
``` python
import xarray as xr
import holoviews as hv
import numpy as np

hv.extension("bokeh")
x = np.arange(0, 3)
y = np.arange(5, 8)
a = np.array([[np.nan, np.nan, 1], [np.nan] * 3, [np.nan] * 3])
b = np.array([[np.nan] * 3, [1, 1, np.nan], [np.nan] * 3])
c = np.array([[np.nan] * 3, [np.nan] * 3, [1, 1, 1]])

ds = (x, y, a, b, c)
im_tuple = hv.ImageStack(ds).opts(title="Tuple")

ds = {"x": x, "y": y, "a": a, "b": b, "c": c}
im_dict = hv.ImageStack(ds).opts(title="Dict")

ds = xr.Dataset(
    {"a": (["x", "y"], a), "b": (["x", "y"], b), "c": (["x", "y"], c)},
    coords={"x": x, "y": y},
)
im_xr = hv.ImageStack(ds).opts(title="xarray")

im_tuple + im_dict + im_xr
```

I get this on the current branch:
image

And this on main:
image

**Comment by Hoxbro** [comment]
This work on the current branch but not on main:
``` python
import xarray as xr
import holoviews as hv
import numpy as np

hv.extension("bokeh")
x = np.arange(0, 3)
y = np.arange(5, 9)
a = np.array([[np.nan, np.nan, 1, np.nan], [np.nan] * 4, [np.nan] * 4])
b = np.array([[np.nan] * 4, [1, 1, np.nan, np.nan], [np.nan] * 4])
c = np.array([[np.nan] * 4, [np.nan] * 4, [1, 1, 1, np.nan]])

ds = {"x": x, "y": y, "a": a, "b": b, "c": c}
im_dict = hv.ImageStack(ds).opts(title="Dict")

ds = xr.Dataset(
    {"a": (["x", "y"], a), "b": (["x", "y"], b), "c": (["x", "y"], c)},
    coords={"x": x, "y": y},
)
im_xr = hv.ImageStack(ds).opts(title="xarray")

im_dict + im_xr
```

image

File: holoviews/element/raster.py [hunk 1]
```diff
@@ -530,21 +540,35 @@ class ImageStack(Image):
...
-        if isinstance(data, np.ndarray) and data.ndim == 3:
-            x = np.arange(data.shape[0])
-            y = np.arange(data.shape[1])
-            data = (x, y, *(data[:, :, n] for n in range(data.shape[2])))
+        _kdims = kdims or self.kdims
+        if isinstance(data, list) and len(data):
+            x = np.arange(data[0].shape[1])
+            y = np.arange(data[0].shape[0])
+            data_tuple = (x, y, *data)
+        elif isinstance(data, dict):
```
Comment (philippjfr) [inline holoviews/element/raster.py:548]: Technically I think a dictionary containing both key dimension coordinate arrays and value dimension arrays should be accepted here.

Comment (philippjfr) [reply holoviews/element/raster.py:548]: e.g. something like: `{'x': np.array([0, 0.5, 1]), 'y': np.array([0, 1]), 'z1': np.random.rand(2, 3), 'z2': np.random.rand(2, 3)}`

And you shouldn't have to do any explicit processing to make this work, the grid interface should already support this.

Comment (Hoxbro) [reply holoviews/element/raster.py:548]: Most of the logic in this if statement is to support this case: `{'z1': np.random.rand(2, 3), 'z2': np.random.rand(2, 3)}`, e.g., no kdims coordinates.

Comment (philippjfr) [reply holoviews/element/raster.py:548]: Yes, but it breaks the case I'm talking about no?

Comment (Hoxbro) [reply holoviews/element/raster.py:548]: It shouldn't. I first try to look up the kdims value in the dictionary, and if those do not exist, I create them. 

image

Maybe I'm just not understanding your question correctly.

Comment (philippjfr) [reply holoviews/element/raster.py:548]: Nope, you are understanding correctly I just didn't read your code correctly :)

File: holoviews/element/raster.py [hunk 2]
```diff
@@ -530,21 +540,35 @@ class ImageStack(Image):
...
+            first = next(v for k, v in data.items() if k not in _kdims)
+            # TODO: Make it work with dict which hv.Dimension key
+            x = data.get(str(_kdims[0]), np.arange(first.shape[1]))
+            y = data.get(str(_kdims[1]), np.arange(first.shape[0]))
```
Comment (philippjfr) [inline holoviews/element/raster.py:552]: Think I'd prefer this for readibility and so we don't pointlessly create arrays:

```suggestion
            xdim, ydim = kdims
            if str(xdim) in data:
                x = data[str(xdim)]
            else:
                x = np.arange(first.shape[1])
            if str(ydim) in data:
                y = data[str(ydim)]
            else:
                y = np.arange(first.shape[1])
```

Comment (Hoxbro) [reply holoviews/element/raster.py:552]: Can we do something like this: `y = data.get(str(_kdims[1])) or np.arange(first.shape[0])`?

Comment (philippjfr) [reply holoviews/element/raster.py:552]: Sure

Comment (philippjfr) [reply holoviews/element/raster.py:552]: Wait, does an array correctly evaluate truthiness?

**Review by philippjfr — APPROVED** [review]

**Pull request merged**